services/api/service.go

```diff
@@ -470,29 +479,54 @@ func (api *RelayAPI) StartServer() (err error) {
 		IdleTimeout:       time.Duration(apiIdleTimeoutMs) * time.Millisecond,
 		MaxHeaderBytes:    apiMaxHeaderBytes,
 	}
-
 	err = api.srv.ListenAndServe()
 	if errors.Is(err, http.ErrServerClosed) {
 		return nil
 	}
 	return err
 }
 
-// StopServer disables sending any bids on getHeader calls, waits a few seconds to catch any remaining getPayload call, and then shuts down the webserver
+func (api *RelayAPI) IsReady() bool {
+	// If server is shutting down, return false
+	if api.srvShutdown.Load() {
+		return false
+	}
+
+	// Proposer API readiness checks
+	if api.opts.ProposerAPI {
+		knownValidatorsUpdated := api.datastore.KnownValidatorsWasUpdated.Load()
+		return knownValidatorsUpdated
+	}
+
+	// Block-builder API readiness checks
+	return true
+}
+
+// StopServer gracefully shuts down the HTTP server:
+// - Stop returning bids
+// - Set ready /readyz to negative status
+// - Wait a bit to allow removal of service from load balancer and draining of requests
 func (api *RelayAPI) StopServer() (err error) {
+	// avoid running this twice. setting srvShutdown to true makes /readyz switch to negative status
+	if wasStopping := api.srvShutdown.Swap(true); wasStopping {
+		return nil
+	}
+
+	// start server shutdown
 	api.log.Info("Stopping server...")
```

Comment:
**issue:**

I think we are misleading the user here a bit.

on one hand we promise that we will respect `apiShutdownWaitDuration` (in seconds).

but on the other we really take more time than that:

- we sleep for `apiShutdownWaitDuration` seconds
- then we wait for in-flighg `getPayload` to complete (yes, most of the time it will be ~0, but that's not guaranteed)
- then we request http-server shutdown (which, again, most of time will be ~0, but - subject to read/write timeouts, and internal request-handling logic - it's also not guaranteed)

what if we use context with timeout to guarantee the shutdown time (and provide logging that'd allow user to diagnose at which step was the shutdown interrupted, if it's terminated because k8s graceful timeout is expired).

Reply:
why would it be better to enforce a strict shutdown time, versus allowing a fully graceful shutdown? I'm tending towards the latter -- not giving time guarantees and allowing the service the time it needs

Reply:
maybe rename `apiShutdownWaitDuration` to `apiShutdownSleepDuration`?

Reply:
> why would it be better to enforce a strict shutdown time

the core of the idea is to:

- have strict timeout (inside the app)
- _plus_, have logging that would indicate to the user - in cases when the timeout was not sufficient - at which stage was the shutdown interrupted

the thing is:  if k8s overruns the graceful shutdown timeout it will just terminate the pod/process and there will be no way to tell at which stage.

if we make our own hard cut-off timeout (and if the user will configure it to be the same as k8s one), then secretly k8s gives 2 more seconds to the pod before it calls off the show.  therefore, we will have enough time to report what was the issue (a.k.a. at which stage we were when the clock ran out).

this will allow the user to reason about what was at the hand.  e.g. was it that some of the `getPayload` calls were taking too long, or was it that some other in-flight requests, or anything else.

the core idea is to make it more troubleshoot-able when things get weird.